### PR TITLE
Move HIR parenting information out of hir_owner

### PR DESCRIPTION
**Pull request by cjgillot**
Split out of #82681.

The parent of a HIR node and its content are currently bundled together, but are rarely used together.
This PR separates both information in two distinct queries for HIR owners.
This reduces incremental invalidation for HIR items that appear within a function body when this body (and the local ids) changes.

### CODE REVIEW

**Comment by cjgillot** [comment]
rustbot did not assign anybody.
r? @petrochenkov

**Comment by petrochenkov** [comment]
I think it's better to r? @michaelwoerister on this.
Also cc @eddyb who had opinions on HIR parenting, IIRC.

File: compiler/rustc_middle/src/query/mod.rs [hunk 1]
```diff
@@ -52,6 +52,15 @@ rustc_queries! {
         desc { |tcx| "HIR owner of `{}`", tcx.def_path_str(key.to_def_id()) }
     }
 
+    /// Gives access to the HIR node's parent for the HIR owner `key`.
+    ///
+    /// This can be conveniently accessed by methods on `tcx.hir()`.
+    /// Avoid calling this query directly.
+    query hir_owner_parent(key: LocalDefId) -> hir::HirId {
+        eval_always
+        desc { |tcx| "HIR parent of `{}`", tcx.def_path_str(key.to_def_id()) }
```
Comment (eddyb) [inline compiler/rustc_middle/src/query/mod.rs:61]: This seems fine to me, I think, if it's not a significant performance loss.

**Comment by michaelwoerister** [comment]
I don't know when I'll find the time to review this. @eddyb, you already looked at it. Would it be a lot of work for you to take over the review?

**Review by eddyb — APPROVED** [review]
r=me (not 100% on the incremental aspects, but it looks like the bulk of the changes account for the reshuffling of the data representation, and shouldn't change behavior outside of increasing incremental precision)

**Comment by Aaron1011** [comment]
It looks like this is a regression on several benchmarks - is there anything that can be done about that?

**Comment by eddyb** [comment]
> It looks like this is a regression on several benchmarks - is there anything that can be done about that?

Oh sheesh, I completely missed that, sorry. Maybe nominate for discussion? I don't have a strong opinion, I just didn't see it.

**Pull request merged**